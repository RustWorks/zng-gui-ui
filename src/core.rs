pub mod animation;
pub mod app;
pub mod context;
pub mod event;
pub mod events;
pub mod focus;
pub mod font;
pub mod frame;
pub mod profiler;
pub mod types;
pub mod var;
pub mod window;

use crate::impl_ui_node;
use context::{LazyStateMap, WidgetContext};
use frame::FrameBuilder;
use types::{LayoutSize, WidgetId};

/// An Ui tree node.
pub trait UiNode: 'static {
    /// Called every time the node is plugged in an Ui tree.
    fn init(&mut self, ctx: &mut WidgetContext);

    /// Called every time the node is unplugged from an Ui tree.
    fn deinit(&mut self, ctx: &mut WidgetContext);

    /// Called every time a low pressure event update happens.
    ///
    /// # Event Pressure
    /// See [update_hp] for more information about event pressure rate.
    fn update(&mut self, ctx: &mut WidgetContext);

    /// Called every time a high pressure event update happens.
    ///
    /// # Event Pressure
    /// Some events occur alot more times then others, for performance reasons this
    /// event source may choose to be propagated in the this hight pressure lane.
    ///
    /// Event sources that are high pressure mention this in their documentation.
    fn update_hp(&mut self, ctx: &mut WidgetContext);

    /// Called every time a layout update is needed.
    ///
    /// # Arguments
    /// * `available_size`: The total available size for the node. Can contain positive infinity to
    /// indicate the parent will accommodate any size.
    ///
    /// # Return
    /// Must return the nodes desired size. Must not contain infinity or NaN.
    fn measure(&mut self, available_size: LayoutSize) -> LayoutSize;

    /// Called every time a layout update is needed, after [measure].
    ///
    /// # Arguments
    /// * `final_size`: The size the parent node reserved for the node. Must reposition its contents
    /// to fit this size. The value does not contain infinity or NaNs.
    fn arrange(&mut self, final_size: LayoutSize);

    /// Called every time a new frame must be rendered.
    ///
    /// # Arguments
    /// * `frame`: Contains the next frame draw instructions.
    fn render(&self, frame: &mut FrameBuilder);

    /// Box this component, unless it is already `Box<dyn UiNode>`.
    fn boxed(self) -> Box<dyn UiNode>
    where
        Self: Sized + 'static,
    {
        Box::new(self)
    }
}

#[impl_ui_node(delegate: self.as_ref(), delegate_mut: self.as_mut())]
impl UiNode for Box<dyn UiNode> {
    fn boxed(self) -> Box<dyn UiNode> {
        self
    }
}

struct Widget<T: UiNode> {
    id: WidgetId,
    state: LazyStateMap,
    child: T,
}

#[impl_ui_node(child)]
impl<T: UiNode> UiNode for Widget<T> {
    fn init(&mut self, ctx: &mut WidgetContext) {
        let child = &mut self.child;
        ctx.widget_context(self.id, &mut self.state, |ctx| child.init(ctx));
    }

    fn deinit(&mut self, ctx: &mut WidgetContext) {
        let child = &mut self.child;
        ctx.widget_context(self.id, &mut self.state, |ctx| child.deinit(ctx));
    }

    fn update(&mut self, ctx: &mut WidgetContext) {
        let child = &mut self.child;
        ctx.widget_context(self.id, &mut self.state, |ctx| child.update(ctx));
    }

    fn update_hp(&mut self, ctx: &mut WidgetContext) {
        let child = &mut self.child;
        ctx.widget_context(self.id, &mut self.state, |ctx| child.update_hp(ctx));
    }

    fn render(&self, frame: &mut FrameBuilder) {
        frame.push_widget(self.id, &self.child);
    }
}

/// Creates a widget bondary.
pub fn widget(id: WidgetId, child: impl UiNode) -> impl UiNode {
    Widget {
        id,
        state: LazyStateMap::default(),
        child,
    }
}
