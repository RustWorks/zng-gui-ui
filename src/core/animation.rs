use crate::core::context::Vars;
use crate::core::var::VarValue;
use std::fmt::Debug;
use std::{ops::Mul, time::Duration};

/// A [variable](crate::core::var::Var) animation.
pub trait VarAnimation<T: VarValue> {
    /// Current animating value.
    fn get_step(&self, vars: &Vars) -> &T;

    /// Update animating value.
    fn update_step(&self, vars: &Vars) -> Option<&T>;

    /// Borrow end value.
    fn get_end(&self) -> &T;

    /// Stops the animation and returns  the end value.
    fn end(self) -> T;
}

/// Easing function value. Mostly between 0.0 and 1.0 but can over/undershot.
#[derive(Debug, Clone, Copy)]
pub struct EasingStep(pub f32);

impl EasingStep {
    /// Gets the scale value.
    #[inline]
    pub fn get(self) -> f32 {
        self.0
    }

    /// Flipped step.
    #[inline]
    pub fn flip(self) -> EasingStep {
        EasingStep(1.0 - self.0)
    }
}

/// Easing function time input. Is a value in the `0.0..=1.0` range.
#[derive(Debug, Clone, Copy)]
pub struct EasingTime(f32);

impl EasingTime {
    /// New easing time from calculated value.
    pub fn from_raw(raw: f32) -> EasingTime {
        EasingTime(raw.max(0.0).min(1.0))
    }

    /// New easing time from total `duration` and `elapsed` time.
    #[inline]
    pub fn new(duration: Duration, elapsed: Duration) -> EasingTime {
        if elapsed < duration {
            EasingTime(duration.as_secs_f32() / elapsed.as_secs_f32())
        } else {
            EasingTime(1.0)
        }
    }

    /// Gets the scale value.
    #[inline]
    pub fn get(self) -> f32 {
        self.0
    }

    /// Inverted time.
    #[inline]
    pub fn reverse(self) -> EasingTime {
        EasingTime(1.0 - self.0)
    }
}

/// A [VarValue] that is the final value of a transition animation and
/// can be multiplied by a [EasingStep] to get an intermediary value of the
/// same type.
pub trait EasingVarValue: VarValue + Mul<EasingStep, Output = Self> {}

impl<T: VarValue + Mul<EasingStep, Output = Self>> EasingVarValue for T {}

impl Mul<EasingStep> for f32 {
    type Output = Self;

    fn mul(self, rhs: EasingStep) -> f32 {
        self * rhs.0
    }
}

impl Mul<EasingStep> for f64 {
    type Output = Self;

    fn mul(self, rhs: EasingStep) -> f64 {
        self * rhs.0 as f64
    }
}

macro_rules! mul_easing_step_for_small_int {
    ($($ty:ty),+) => {$(
        impl Mul<EasingStep> for $ty {
            type Output = Self;

            #[inline]
            fn mul(self, rhs: EasingStep) -> Self {
                (self as f32 * rhs.0) as Self
            }
        }
    )+};
}

mul_easing_step_for_small_int!(u8, i8, u16, i16, u32, i32);

macro_rules! mul_easing_step_for_int  {
    ($($ty:ty),+) => {$(
        impl Mul<EasingStep> for $ty {
            type Output = Self;

            #[inline]
            fn mul(self, rhs: EasingStep) -> Self {
                (self as f64 * rhs.0 as f64) as Self
            }
        }
    )+};
}

mul_easing_step_for_int!(usize, isize, u64, i64, u128, i128);

/// Common easing functions.
///
/// See also: [`EasingFn`](EasingFn).
pub mod easing {
    use super::{EasingStep, EasingTime};
    use std::f32::consts::*;

    /// Applies the `ease_fn`.
    #[inline]
    pub fn ease_in(ease_fn: impl FnOnce(EasingTime) -> EasingStep, time: EasingTime) -> EasingStep {
        ease_fn(time)
    }

    /// Applies the `ease_fn` in reverse and flipped.
    #[inline]
    pub fn ease_out(ease_fn: impl FnOnce(EasingTime) -> EasingStep, time: EasingTime) -> EasingStep {
        ease_fn(time.reverse()).flip()
    }

    /// Applies `ease_in` for the first half then `[ease_out]` scaled to fit a single duration (1.0).
    pub fn ease_in_out(ease_fn: impl FnOnce(EasingTime) -> EasingStep, time: EasingTime) -> EasingStep {
        let time = EasingTime(time.get() * 2.0);
        let step = if time.get() < 1.0 {
            ease_in(ease_fn, time)
        } else {
            ease_out(ease_fn, time)
        };
        EasingStep(step.get() * 0.5)
    }

    /// Simple linear transition, no easing, no acceleration.
    #[inline]
    pub fn linear(time: EasingTime) -> EasingStep {
        EasingStep(time.get())
    }

    #[inline]
    pub fn quad(time: EasingTime) -> EasingStep {
        let t = time.get();
        EasingStep(t * t)
    }

    #[inline]
    pub fn cubic(time: EasingTime) -> EasingStep {
        let t = time.get();
        EasingStep(t * t * t)
    }

    #[inline]
    pub fn quart(time: EasingTime) -> EasingStep {
        let t = time.get();
        EasingStep(t * t * t * t)
    }

    #[inline]
    pub fn quint(time: EasingTime) -> EasingStep {
        let t = time.get();
        EasingStep(t * t * t * t * t)
    }

    #[inline]
    pub fn sine(time: EasingTime) -> EasingStep {
        let t = time.get();
        EasingStep(1.0 - (t * FRAC_PI_2 * (1.0 - t)).sin())
    }

    #[inline]
    pub fn expo(time: EasingTime) -> EasingStep {
        let t = time.get();
        EasingStep((10.0 * (t - 1.0)).powf(2.0))
    }

    #[inline]
    pub fn circ(time: EasingTime) -> EasingStep {
        EasingStep(1.0 - (1.0 - time.get()).sqrt())
    }

    #[inline]
    pub fn back(time: EasingTime) -> EasingStep {
        let t = time.get();
        EasingStep(t * t * (2.70158 * t - 1.70158))
    }

    #[inline]
    pub fn elastic(time: EasingTime) -> EasingStep {
        let t = time.get();
        let t2 = t * t;
        EasingStep(t2 * t2 * (t * PI * 4.5).sin())
    }

    #[inline]
    pub fn bounce(time: EasingTime) -> EasingStep {
        let t = time.get();
        EasingStep((6.0 * (t - 1.0)).powf(2.0) * (t * PI * 3.5).sin().abs())
    }
}

/// Common easing functions as an enum.
#[derive(Debug, Clone, Copy)]
pub enum EasingFn {
    Linear,
    Sine,
    Quad,
    Cubic,
    Quart,
    Quint,
    Expo,
    Circ,
    Back,
    Elastic,
    Bounce,
}

impl EasingFn {
    #[inline]
    pub fn ease_in(self, time: EasingTime) -> EasingStep {
        match self {
            EasingFn::Linear => easing::linear(time),
            EasingFn::Sine => easing::sine(time),
            EasingFn::Quad => easing::quad(time),
            EasingFn::Cubic => easing::cubic(time),
            EasingFn::Quart => easing::quad(time),
            EasingFn::Quint => easing::quint(time),
            EasingFn::Expo => easing::expo(time),
            EasingFn::Circ => easing::circ(time),
            EasingFn::Back => easing::back(time),
            EasingFn::Elastic => easing::elastic(time),
            EasingFn::Bounce => easing::bounce(time),
        }
    }

    #[inline]
    pub fn ease_out(self, time: EasingTime) -> EasingStep {
        easing::ease_out(|t| self.ease_in(t), time)
    }

    #[inline]
    pub fn ease_in_out(self, time: EasingTime) -> EasingStep {
        easing::ease_in_out(|t| self.ease_in(t), time)
    }
}
