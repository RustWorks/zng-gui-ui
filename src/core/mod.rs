#[macro_use]
mod macros;

mod focus;
mod font;
mod hittest;
mod keyboard;
mod mouse;
mod next_frame;
mod next_update;
mod ui;
mod ui_values;
mod window;

pub use focus::*;
pub use font::*;
pub use glutin::event::{ElementState, ModifiersState, MouseButton, ScanCode, VirtualKeyCode};
pub use glutin::window::CursorIcon;
pub use hittest::*;
pub use keyboard::*;
pub use mouse::*;
pub use next_frame::*;
pub use next_update::*;
pub use ui::*;
pub use ui_values::*;
pub use webrender::api::{
    ColorF, FontInstanceKey, FontKey, GradientStop, LayoutPoint, LayoutRect, LayoutSideOffsets, LayoutSize,
};
pub use webrender::euclid::{point2, size2};
pub use window::*;
