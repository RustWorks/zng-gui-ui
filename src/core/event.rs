//! App event API.

use crate::core::context::{Events, WidgetContext};
use std::cell::{Cell, UnsafeCell};
use std::fmt::Debug;
use std::rc::Rc;
use std::time::Instant;

/// [`Event`] arguments.
pub trait EventArgs: Debug + Clone + 'static {
    /// Gets the instant this event happen.
    fn timestamp(&self) -> Instant;
    /// If this event arguments is relevant to the widget context.
    fn concerns_widget(&self, _: &mut WidgetContext) -> bool;
}

/// [`Event`] arguments that can be canceled.
pub trait CancelableEventArgs: EventArgs {
    /// If the originating action must be canceled.
    fn cancel_requested(&self) -> bool;
    /// Cancel the originating action.
    fn cancel(&self);
}

/// Identifies an event type.
pub trait Event: 'static {
    /// Event arguments type.
    type Args: EventArgs;
    /// If the event is updated in the high-pressure lane.
    const IS_HIGH_PRESSURE: bool = false;

    /// New event emitter.
    fn emitter() -> EventEmitter<Self::Args> {
        EventEmitter::new(Self::IS_HIGH_PRESSURE)
    }

    /// New event listener that never updates.
    fn never() -> EventListener<Self::Args> {
        EventListener::never(Self::IS_HIGH_PRESSURE)
    }
}

/// Identifies an event type for an action that can be canceled.
///
/// # Auto-Implemented
///
/// This trait is auto-implemented for all events with cancellable arguments.
pub trait CancelableEvent: Event + 'static {
    /// Cancelable event arguments type.
    type CancelableArgs: CancelableEventArgs;
}
impl<A: CancelableEventArgs, E: Event<Args = A>> CancelableEvent for E {
    type CancelableArgs = A;
}

struct EventChannelInner<T> {
    data: UnsafeCell<Vec<T>>,
    listener_count: Cell<usize>,
    is_high_pressure: bool,
}

struct EventChannel<T: 'static> {
    r: Rc<EventChannelInner<T>>,
}
impl<T: 'static> Clone for EventChannel<T> {
    fn clone(&self) -> Self {
        EventChannel { r: Rc::clone(&self.r) }
    }
}
impl<T: 'static> EventChannel<T> {
    pub(crate) fn notify(self, new_update: T, _assert_events_not_borrowed: &mut Events, cleanup: &mut Vec<Box<dyn FnOnce()>>) {
        // SAFETY: This is safe because borrows are bound to the `Events` instance
        // so if we have a mutable reference to it no event value is borrowed.
        let data = unsafe { &mut *self.r.data.get() };
        data.push(new_update);

        if data.len() == 1 {
            // register for cleanup once
            cleanup.push(Box::new(move || {
                unsafe { &mut *self.r.data.get() }.clear();
            }))
        }
    }

    /// Gets a reference to the updates that happened in between calls of [`UiNode::update`](crate::core::UiNode::update).
    pub fn updates<'a>(&'a self, _events: &'a Events) -> &'a [T] {
        // SAFETY: This is safe because we are bounding the value lifetime with
        // the `Events` lifetime and we require a mutable reference to `Events` to
        // modify the value.
        unsafe { &*self.r.data.get() }.as_ref()
    }

    /// If this update is notified using the [`UiNode::update_hp`](crate::core::UiNode::update_hp) method.
    pub fn is_high_pressure(&self) -> bool {
        self.r.is_high_pressure
    }

    pub fn listener_count(&self) -> usize {
        self.r.listener_count.get()
    }

    pub fn has_listeners(&self) -> bool {
        self.listener_count() > 0
    }

    pub fn on_new_listener(&self) {
        self.r.listener_count.set(self.r.listener_count.get() + 1)
    }

    pub fn on_drop_listener(&self) {
        self.r.listener_count.set(self.r.listener_count.get() - 1)
    }
}

/// Read-only reference to an event channel.
pub struct EventListener<T: 'static> {
    chan: EventChannel<T>,
}
impl<T: 'static> Clone for EventListener<T> {
    fn clone(&self) -> Self {
        EventListener::new(self.chan.clone())
    }
}
impl<T: 'static> EventListener<T> {
    fn new(chan: EventChannel<T>) -> Self {
        chan.on_new_listener();
        EventListener { chan }
    }

    fn never(is_high_pressure: bool) -> Self {
        EventEmitter::new(is_high_pressure).into_listener()
    }

    /// Gets a reference to the updates that happened in between calls of [`UiNode::update`](crate::core::UiNode::update).
    pub fn updates<'a>(&'a self, events: &'a Events) -> &'a [T] {
        self.chan.updates(events)
    }

    /// If [`updates`](EventListener::updates) is not empty.
    pub fn has_updates<'a>(&'a self, events: &'a Events) -> bool {
        !self.updates(events).is_empty()
    }

    /// If this update is notified using the [`UiNode::update_hp`](crate::core::UiNode::update_hp) method.
    pub fn is_high_pressure(&self) -> bool {
        self.chan.is_high_pressure()
    }
}

impl<T: 'static> Drop for EventListener<T> {
    fn drop(&mut self) {
        self.chan.on_drop_listener();
    }
}

/// Read-write reference to an event channel.
pub struct EventEmitter<T: 'static> {
    chan: EventChannel<T>,
}
impl<T: 'static> Clone for EventEmitter<T> {
    fn clone(&self) -> Self {
        EventEmitter { chan: self.chan.clone() }
    }
}
impl<T: 'static> EventEmitter<T> {
    fn new(is_high_pressure: bool) -> Self {
        EventEmitter {
            chan: EventChannel {
                r: Rc::new(EventChannelInner {
                    data: UnsafeCell::default(),
                    listener_count: Cell::new(0),
                    is_high_pressure,
                }),
            },
        }
    }

    /// New emitter for a service request response.
    ///
    /// The emitter is expected to update at maximum only once so it is not high-pressure.
    pub fn response() -> Self {
        Self::new(false)
    }

    /// Number of listener to this event emitter.
    pub fn listener_count(&self) -> usize {
        self.chan.listener_count()
    }

    /// If this event emitter has any listeners.
    pub fn has_listeners(&self) -> bool {
        self.chan.has_listeners()
    }

    /// Gets a reference to the updates that happened in between calls of [`UiNode::update`](crate::core::UiNode::update).
    pub fn updates<'a>(&'a self, events: &'a Events) -> &'a [T] {
        self.chan.updates(events)
    }

    /// If [`updates`](EventEmitter::updates) is not empty.
    pub fn has_updates<'a>(&'a self, events: &'a Events) -> bool {
        !self.updates(events).is_empty()
    }

    /// If this event is notified using the [`UiNode::update_hp`](crate::core::UiNode::update_hp) method.
    pub fn is_high_pressure(&self) -> bool {
        self.chan.is_high_pressure()
    }

    /// Gets a new event listener linked with this emitter.
    pub fn listener(&self) -> EventListener<T> {
        EventListener::new(self.chan.clone())
    }

    /// Converts this emitter instance into a listener.
    pub fn into_listener(self) -> EventListener<T> {
        EventListener::new(self.chan)
    }

    pub(crate) fn notify(self, new_update: T, assert_events_not_borrowed: &mut Events, cleanup: &mut Vec<Box<dyn FnOnce()>>) {
        self.chan.notify(new_update, assert_events_not_borrowed, cleanup);
    }
}

pub use zero_ui_macros::{cancelable_event_args, event, event_args, event_hp};
