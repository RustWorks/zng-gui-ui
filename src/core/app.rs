use crate::core::event::EventEmitter;
use crate::core::event::EventListener;
use crate::core::{
    context::*, event::CancelableEventArgs, focus::FocusManager, font::FontManager, gesture::GestureEvents, keyboard::KeyboardEvents,
    mouse::MouseEvents, types::*, window::WindowManager,
};
use glutin::event::Event as GEvent;
use glutin::event_loop::{ControlFlow, EventLoop};
use std::any::{type_name, TypeId};
use std::mem;

/// An [App] extension.
pub trait AppExtension: 'static {
    /// Type id of this extension.
    #[inline]
    fn id(&self) -> TypeId {
        TypeId::of::<Self>()
    }

    /// If this extension is the `app_extension_id` or dispatches to it.
    #[inline]
    fn is_or_contain(&self, app_extension_id: TypeId) -> bool {
        self.id() == app_extension_id
    }

    /// Initializes this extension.
    #[inline]
    fn init(&mut self, _ctx: &mut AppInitContext) {}

    /// Called when the OS sends an event to a device.
    #[inline]
    fn on_device_event(&mut self, _device_id: DeviceId, _event: &DeviceEvent, _ctx: &mut AppContext) {}

    /// Called when the OS sends an event to a window.
    #[inline]
    fn on_window_event(&mut self, _window_id: WindowId, _event: &WindowEvent, _ctx: &mut AppContext) {}

    /// Called when a new frame is ready to be presented.
    #[inline]
    fn on_new_frame_ready(&mut self, _window_id: WindowId, _ctx: &mut AppContext) {}

    /// Called every update after the Ui update.
    #[inline]
    fn update(&mut self, _update: UpdateRequest, _ctx: &mut AppContext) {}

    /// Called after every sequence of updates if display update was requested.
    #[inline]
    fn update_display(&mut self, _update: UpdateDisplayRequest, _ctx: &mut AppContext) {}

    /// Called when the OS sends a request for re-drawing the last frame.
    #[inline]
    fn on_redraw_requested(&mut self, _window_id: WindowId, _ctx: &mut AppContext) {}

    /// Called when a shutdown was requested.
    #[inline]
    fn on_shutdown_requested(&mut self, _args: &ShutdownRequestedArgs, _ctx: &mut AppContext) {}

    /// Called when the application is shutting down.
    ///
    /// Update requests generated during this call are ignored.
    #[inline]
    fn deinit(&mut self, _ctx: &mut AppContext) {}
}

cancelable_event_args! {
    /// Arguments for `on_shutdown_requested`.
    pub struct ShutdownRequestedArgs {
        ..
        /// Always true.
        fn concerns_widget(&self, _ctx: &mut WidgetContext) -> bool {
            true
        }
    }
}

impl AppExtension for () {}

impl<A: AppExtension, B: AppExtension> AppExtension for (A, B) {
    #[inline]
    fn init(&mut self, ctx: &mut AppInitContext) {
        self.0.init(ctx);
        self.1.init(ctx);
    }

    #[inline]
    fn is_or_contain(&self, app_extension_id: TypeId) -> bool {
        self.0.is_or_contain(app_extension_id) || self.1.is_or_contain(app_extension_id) || self.id() == app_extension_id
    }

    #[inline]
    fn on_device_event(&mut self, device_id: DeviceId, event: &DeviceEvent, ctx: &mut AppContext) {
        self.0.on_device_event(device_id, event, ctx);
        self.1.on_device_event(device_id, event, ctx);
    }

    #[inline]
    fn on_window_event(&mut self, window_id: WindowId, event: &WindowEvent, ctx: &mut AppContext) {
        self.0.on_window_event(window_id, event, ctx);
        self.1.on_window_event(window_id, event, ctx);
    }

    #[inline]
    fn on_new_frame_ready(&mut self, window_id: WindowId, ctx: &mut AppContext) {
        self.0.on_new_frame_ready(window_id, ctx);
        self.1.on_new_frame_ready(window_id, ctx);
    }

    #[inline]
    fn update(&mut self, update: UpdateRequest, ctx: &mut AppContext) {
        self.0.update(update, ctx);
        self.1.update(update, ctx);
    }

    #[inline]
    fn update_display(&mut self, update: UpdateDisplayRequest, ctx: &mut AppContext) {
        self.0.update_display(update, ctx);
        self.1.update_display(update, ctx);
    }

    #[inline]
    fn on_redraw_requested(&mut self, window_id: WindowId, ctx: &mut AppContext) {
        self.0.on_redraw_requested(window_id, ctx);
        self.1.on_redraw_requested(window_id, ctx);
    }

    #[inline]
    fn on_shutdown_requested(&mut self, args: &ShutdownRequestedArgs, ctx: &mut AppContext) {
        self.0.on_shutdown_requested(args, ctx);
        self.1.on_shutdown_requested(args, ctx);
    }

    #[inline]
    fn deinit(&mut self, ctx: &mut AppContext) {
        self.0.deinit(ctx);
        self.1.deinit(ctx);
    }
}

/// Defines and runs an application.
pub struct App;

impl App {
    /// Application without any extension.
    #[inline]
    pub fn empty() -> AppExtended<()> {
        AppExtended { extensions: () }
    }

    /// Application with default extensions.
    ///
    /// # Extensions
    ///
    /// Extensions included.
    ///
    /// * [MouseEvents]
    /// * [KeyboardEvents]
    /// * [GestureEvents]
    /// * [WindowManager]
    /// * [FontManager]
    /// * [FocusManager]
    #[inline]
    pub fn default() -> AppExtended<impl AppExtension> {
        App::empty()
            .extend(MouseEvents::default())
            .extend(KeyboardEvents::default())
            .extend(GestureEvents::default())
            .extend(WindowManager::default())
            .extend(FontManager::default())
            .extend(FocusManager::default())
    }
}

/// Application with extensions.
pub struct AppExtended<E: AppExtension> {
    extensions: E,
}

pub struct ShutDownCancelled;

/// Service for managing the application process.
pub struct AppProcess {
    shutdown_requests: Vec<EventEmitter<ShutDownCancelled>>,
    update_notifier: UpdateNotifier,
}
impl AppService for AppProcess {}
impl AppProcess {
    pub fn new(update_notifier: UpdateNotifier) -> Self {
        AppProcess {
            shutdown_requests: Vec::new(),
            update_notifier,
        }
    }

    /// Register a request for process shutdown in the next update.
    ///
    /// Returns an event listener that is updated once with the unit value [`ShutDownCancelled`](ShutDownCancelled)
    /// if the shutdown operation is cancelled.
    pub fn shutdown(&mut self) -> EventListener<ShutDownCancelled> {
        let emitter = EventEmitter::new(false);
        self.shutdown_requests.push(emitter.clone());
        self.update_notifier.push_update();
        emitter.into_listener()
    }

    fn take_requests(&mut self) -> Vec<EventEmitter<ShutDownCancelled>> {
        mem::take(&mut self.shutdown_requests)
    }
}
///Returns if should shutdown
fn shutdown(shutdown_requests: Vec<EventEmitter<ShutDownCancelled>>, ctx: &mut AppContext, ext: &mut impl AppExtension) -> bool {
    if shutdown_requests.is_empty() {
        return false;
    }
    let args = ShutdownRequestedArgs::now();
    ext.on_shutdown_requested(&args, ctx);
    if args.cancel_requested() {
        for c in shutdown_requests {
            ctx.updates.push_notify(c, ShutDownCancelled)
        }
    }
    !args.cancel_requested()
}

impl<E: AppExtension> AppExtended<E> {
    /// Gets if the application is already extended with the extension type.
    #[inline]
    pub fn extended_with<F: AppExtension>(&self) -> bool {
        self.extensions.is_or_contain(TypeId::of::<F>())
    }

    /// Includes an application extension.
    ///
    /// # Panics
    /// * `"app already extended with `{}`"` when the app is already [`extended_with`](AppExtended::extended_with) the
    /// extension type.
    #[inline]
    pub fn extend<F: AppExtension>(self, extension: F) -> AppExtended<impl AppExtension> {
        if self.extended_with::<F>() {
            panic!("app already extended with `{}`", type_name::<F>())
        }
        AppExtended {
            extensions: (self.extensions, extension),
        }
    }

    /// Runs the application event loop calling `start` once at the beginning.
    #[inline]
    pub fn run(self, start: impl FnOnce(&mut AppContext)) -> ! {
        #[cfg(feature = "app_profiler")]
        crate::core::profiler::register_thread_with_profiler();

        profile_scope!("app::run");

        let event_loop = EventLoop::with_user_event();

        let mut extensions = self.extensions;

        let mut owned_ctx = OwnedAppContext::instance(event_loop.create_proxy());

        let mut init_ctx = owned_ctx.borrow_init();
        init_ctx.services.register(AppProcess::new(init_ctx.updates.notifier().clone()));
        extensions.init(&mut init_ctx);

        let mut in_sequence = false;
        let mut sequence_update = UpdateDisplayRequest::None;

        start(&mut owned_ctx.borrow(&event_loop));

        event_loop.run(move |event, event_loop, control_flow| {
            profile_scope!("app::event");

            *control_flow = ControlFlow::Wait;

            let mut event_update = UpdateRequest::default();
            match event {
                GEvent::NewEvents(_) => {
                    in_sequence = true;
                }

                GEvent::WindowEvent { window_id, event } => {
                    profile_scope!("app::on_window_event");
                    extensions.on_window_event(window_id, &event, &mut owned_ctx.borrow(event_loop));
                }
                GEvent::UserEvent(AppEvent::NewFrameReady(window_id)) => {
                    profile_scope!("app::on_new_frame_ready");
                    extensions.on_new_frame_ready(window_id, &mut owned_ctx.borrow(event_loop));
                }
                GEvent::UserEvent(AppEvent::Update) => {
                    event_update = owned_ctx.take_request();
                }
                GEvent::DeviceEvent { device_id, event } => {
                    profile_scope!("app::on_device_event");
                    extensions.on_device_event(device_id, &event, &mut owned_ctx.borrow(event_loop));
                }

                GEvent::MainEventsCleared => {
                    in_sequence = false;
                }

                GEvent::RedrawRequested(window_id) => {
                    profile_scope!("app::on_redraw_requested");
                    extensions.on_redraw_requested(window_id, &mut owned_ctx.borrow(event_loop))
                }

                #[cfg(feature = "app_profiler")]
                GEvent::LoopDestroyed => {
                    crate::core::profiler::write_profile("app_profile.json", true);
                }

                _ => {}
            }

            let mut limit = 100_000;
            loop {
                let (mut update, display) = owned_ctx.apply_updates();
                update |= mem::take(&mut event_update);
                sequence_update |= display;

                if update.update || update.update_hp {
                    profile_scope!("app::update");
                    let mut ctx = owned_ctx.borrow(event_loop);
                    let shutdown_requests = ctx.services.req::<AppProcess>().take_requests();
                    if shutdown(shutdown_requests, &mut ctx, &mut extensions) {
                        *control_flow = ControlFlow::Exit;
                        return;
                    }
                    extensions.update(update, &mut ctx);
                } else {
                    break;
                }

                limit -= 1;
                if limit == 0 {
                    panic!("immediate update loop reached limit of `100_000` repeats")
                }
            }

            if !in_sequence && sequence_update.is_some() {
                profile_scope!("app::update_display");
                extensions.update_display(sequence_update, &mut owned_ctx.borrow(event_loop));
                sequence_update = UpdateDisplayRequest::None;
            }
        })
    }
}

#[derive(Debug)]
pub enum AppEvent {
    NewFrameReady(WindowId),
    Update,
}
