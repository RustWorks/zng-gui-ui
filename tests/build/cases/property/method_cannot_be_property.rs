use zero_ui::core::{
    property,
    var::IntoVar,
    widget_instance::{NilUiNode, UiNode},
};

struct Foo;
impl Foo {
    #[property(context)]
    pub fn self_method1(self, input: impl IntoVar<bool>) -> impl UiNode {
        NilUiNode
    }

    #[property(context)]
    pub fn self_method2(self: Box<Self>, input: impl IntoVar<bool>) -> impl UiNode {
        NilUiNode
    }
}

fn main() {
    let _mtd_was_not_removed = Foo.self_method1(true);
    let _mtd_was_not_removed = Box::new(Foo).self_method2(true);
}
