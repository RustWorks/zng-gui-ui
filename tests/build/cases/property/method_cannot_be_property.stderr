error: methods cannot be properties
  --> cases/property/method_cannot_be_property.rs:10:25
   |
10 |     pub fn self_method1(self, input: impl IntoVar<bool>) -> impl UiNode {
   |                         ^^^^

error: methods cannot be properties
  --> cases/property/method_cannot_be_property.rs:16:25
   |
16 |     pub fn self_method2(self: Box<Self>, input: impl IntoVar<bool>) -> impl UiNode {
   |                         ^^^^
